src/main/scala/org/broadinstitute/hail/Utils.scala

```diff
@@ -447,35 +450,52 @@ class RichRDDByteArray(val r: RDD[Array[Byte]]) extends AnyVal {
   }
 }
 
-class RichPairRDD[K, V](val r: RDD[(K, V)]) extends AnyVal {
+class RichPairRDD[K, V](val rdd: RDD[(K, V)]) extends AnyVal {
+
+  def forall(p: ((K, V)) => Boolean)(implicit kct: ClassTag[K], vct: ClassTag[V]): Boolean = rdd.map(p).fold(true)(_ && _)
 
-  def forall(p: ((K, V)) => Boolean)(implicit kct: ClassTag[K], vct: ClassTag[V]): Boolean = r.map(p).fold(true)(_ && _)
+  def exists(p: ((K, V)) => Boolean)(implicit kct: ClassTag[K], vct: ClassTag[V]): Boolean = rdd.map(p).fold(false)(_ || _)
 
-  def exists(p: ((K, V)) => Boolean)(implicit kct: ClassTag[K], vct: ClassTag[V]): Boolean = r.map(p).fold(false)(_ || _)
+  def mapValuesWithKey[W](f: (K, V) => W): RDD[(K, W)] = rdd.mapPartitions(_.map { case (k, v) => (k, f(k, v)) },
+    preservesPartitioning = true)
+
+  def flatMapValuesWithKey[W](f: (K, V) => TraversableOnce[W]): RDD[(K, W)] = rdd.mapPartitions(_.flatMap { case (k, v) =>
+    f(k, v).map(w => (k, w))
+  }, preservesPartitioning = true)
 
   def spanByKey()(implicit kct: ClassTag[K], vct: ClassTag[V]): RDD[(K, Iterable[V])] =
-    r.mapPartitions(p => new SpanningIterator(p))
+    rdd.mapPartitions(p => new SpanningIterator(p))
 
   def leftOuterJoinDistinct[W](other: RDD[(K, W)])
-    (implicit kt: ClassTag[K], vt: ClassTag[V], ord: Ordering[K] = null): RDD[(K, (V, Option[W]))] = leftOuterJoinDistinct(other, defaultPartitioner(r, other))
+    (implicit kt: ClassTag[K], vt: ClassTag[V], ord: Ordering[K] = null): RDD[(K, (V, Option[W]))] = leftOuterJoinDistinct(other, defaultPartitioner(rdd, other))
 
   def leftOuterJoinDistinct[W](other: RDD[(K, W)], partitioner: Partitioner)
     (implicit kt: ClassTag[K], vt: ClassTag[V], ord: Ordering[K] = null) = {
-    r.cogroup(other, partitioner).flatMapValues { pair =>
+    rdd.cogroup(other, partitioner).flatMapValues { pair =>
       val w = pair._2.headOption
       pair._1.map((_, w))
     }
   }
 
   def joinDistinct[W](other: RDD[(K, W)])
-    (implicit kt: ClassTag[K], vt: ClassTag[V], ord: Ordering[K] = null): RDD[(K, (V, W))] = joinDistinct(other, defaultPartitioner(r, other))
+    (implicit kt: ClassTag[K], vt: ClassTag[V], ord: Ordering[K] = null): RDD[(K, (V, W))] = joinDistinct(other, defaultPartitioner(rdd, other))
 
   def joinDistinct[W](other: RDD[(K, W)], partitioner: Partitioner)
     (implicit kt: ClassTag[K], vt: ClassTag[V], ord: Ordering[K] = null) = {
-    r.cogroup(other, partitioner).flatMapValues { pair =>
+    rdd.cogroup(other, partitioner).flatMapValues { pair =>
       for (v <- pair._1.iterator; w <- pair._2.iterator.take(1)) yield (v, w)
     }
   }
+
+  def orderedLeftJoinDistinct[T, V2](other: RDD[(K, V2)])(implicit ev: (K) => T, tOrd: Ordering[T], kOrd1: Ordering[K],
```

Comment:
Again, isn't this just a specialization of the transformed version with K1 = k2?  Seems like you can delete.